File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    int kthSmallest(TreeNode *root, int k)
    {
        vector<int> nodes;
        dfs(nodes, root);
        return nodes[k - 1];
    }

private:
    void dfs(vector<int> &nodes, TreeNode *root)
    {
        if (root->left == nullptr)
        {
            nodes.push_back(root->val);
            if (root->right != nullptr)
                dfs(nodes, root->right);
            return;
        }
        dfs(nodes, root->left);
        nodes.push_back(root->val);
        if (root->right != nullptr)
            dfs(nodes, root->right);
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        diameter = 0
    
        def height(node):
            nonlocal diameter
            if not node:
                return 0
            leftHeight = height(node.left)
            rightHeight = height(node.right)
            diameter = max(diameter, leftHeight + rightHeight)
            return max(leftHeight, rightHeight) + 1
        height(root)
        return diameter